src/headless/converse-muc.js

```diff
@@ -2361,7 +2361,14 @@ converse.plugins.add('converse-muc', {
                 const body = message.get('message');
                 if (!body) { return; }
                 if (u.isNewMessage(message) && this.isHidden()) {
-                    const settings = {'num_unread_general': this.get('num_unread_general') + 1};
+                    let first_unread = this.get('first_unread');
+
+                    if (this.get('num_unread_general') == 0) {
+                      if (first_unread) first_unread.set("first_unread", false);
+                      message.set("first_unread", true);
+                      first_unread = message;
+                    }
+                    const settings = {'first_unread': first_unread, 'num_unread_general': this.get('num_unread_general') + 1};
```

Comment:
Looks to me like you're saving a message model as the value of an attribute on a ChatBox or ChatRoom instance.

AFAIK this won't work because when the data is restored from localStorage, you won't get the original Message instance back, you'll get a map back.

Besides that, all the model's attributes gets JSON serialized and stored in localStorage, which is wasteful and problematic because localStorage is limited in size.

You'll need to use (and store) something like the `id` value of the first unread message and then use that to look up the Message instance when you need it.

Reply:
OK. thanks for the review. I will store the message id instead as suggested.